devito/operator/operator.py

```diff
@@ -640,6 +642,11 @@ def _prepare_arguments(self, autotune=None, **kwargs):
         for d in reversed(toposort):
             args.update(d._arg_values(self._dspace[d], grid, **kwargs))
 
+        # Process SubDimensionThicknesses
+        for p in self.parameters:
```

Comment:
I think this should be triggered from within `SubDimension._arg_values` itself

Reply:
No, the issue is that the `SubDimension` isn't necessarily available to the operator. You can craft legal `Operator`s which only contain the thickness, not the parent subdimension, hence the reworking in this PR

Reply:
Why isn't it caught by `self.input` 's default line 487

Reply:
Assuming "it" means the subdomain, you can craft a legal operator as in the MFE at the top of this PR where the Operator never sees the SubDimension, only its thickness symbols

Reply:
I don't mean the subdomain no, I mean the thickness, you're saying you can create an operator where the Thickness (p here) is in op.parameters but is not in op.inputs?

If it's the case, i.e if it is *not* in inputs, then it should not need to be in args since it's not an input.
If it's not the case, the line 487 processes all inputs and should cover that case already

Reply:
What do you mean? Several things that need `_arg_values` calling on them are in `self.parameters` and not `self.input` (`x_m` and `x_M` for example).  The argument I'm making is the thickness `ltkn` can appear in an operator without its parent subdimension `ix`

Reply:
I **think** if you add `is_Input = True` to `class Thickness`, then you don't need this extra loop

for example  https://github.com/devitocodes/devito/blob/master/devito/types/constant.py#L42

Reply:
I tried it, and this is not the case

Reply:
How is that possible? It's clearly in parameters since you added this so if you add is_Input and it doesn't catch it then something is broken

Reply:
There's no grid in the kwargs at the point where `_arg_values` is called on the inputs, which makes it no-op. It also needs to be done before objects get `_arg_values` called on them iirc, hence the current position of this loop.

You could reorder `_prepare_arguments` to process the grid earlier, but my functions-on-subdomains branch has a bunch of reworking here and fixing the merge conflicts will be a pain, so best to stick a pin in it and fix down the line imo.

Reply:
Can you open an issue about it then

Reply:
I also don't understand this thing honestly. We need an issue and ideally a comment on top of this loop explaining what's going on and what should actually happen instead. You can do it in another branch as it's a minor thing